sdk/python/src/dagger/mod/_module.py

```diff
@@ -140,43 +263,217 @@ async def _call(self, name: str) -> tuple[str, int]:
+            if not callable(func):
+                msg = f"Expected a callable, got {type(func)}."
+                raise UserError(msg)
+
+            resolver = FunctionResolver(
+                original_name=func.__name__,
+                name=name or func.__name__,
+                wrapped_func=func,
+                doc=doc or get_doc(func) or "",
+                # This will be filled later with @object_type
+                # because it's not bound to a class yet.
+                origin=None,
+            )
+
+            # This is used by `object_type` to find the resolver
+            # for updating the origin.
+            func.__dagger_resolver__ = resolver
+
+            self.add_resolver(resolver)
+
             return func
 
         return wrapper(func) if func else wrapper
+
+    @overload
+    @dataclass_transform(
+        kw_only_default=True,
+        field_specifiers=(function, dataclasses.field, dataclasses.Field),
+    )
+    def object_type(self, cls: T) -> T:
+        ...
+
+    @overload
+    @dataclass_transform(
+        kw_only_default=True,
+        field_specifiers=(function, dataclasses.field, dataclasses.Field),
+    )
+    def object_type(self) -> Callable[[T], T]:
+        ...
+
+    def object_type(self, cls: T | None = None) -> T | Callable[[T], T]:
+        """Exposes a Python class as a :py:class:`dagger.ObjectTypeDef`.
+
+        Used with :py:meth:`field` and :py:meth:`function` to expose
+        the object's members.
+
+        Example usage:
+
+        >>> @object_type
+        >>> class Foo:
+        >>>     @function
+        >>>     def bar(self) -> str:
+        >>>         return "foobar"
+        """
+
+        def wrapper(cls: T) -> T:
+            if not inspect.isclass(cls):
+                msg = f"Expected a class, got {type(cls)}"
+                raise UserError(msg)
+
+            wrapped = dataclass(kw_only=True)(cls)
+            return self._process_type(wrapped)
+
+        return wrapper(cls) if cls else wrapper
+
+    def _process_type(self, cls: T) -> T:
+        types = typing.get_type_hints(cls)
+
+        for field in dataclasses.fields(cls):
+            field_def: FieldDefinition | None
+            if field_def := field.metadata.get(FIELD_DEF_KEY, None):
+                r = FieldResolver(
+                    name=field_def.name or field.name,
+                    original_name=field.name,
+                    doc=get_doc(field.type),
+                    type_annotation=types[field.name],
+                    is_optional=field_def.optional,
+                    origin=cls,
+                )
+                self.add_resolver(r)
+
+        for _, member in inspect.getmembers(cls):
+            resolver: FunctionResolver | None
+            if resolver := getattr(member, "__dagger_resolver__", None):
+                resolver.origin = cls
+
+        cls.__dagger_type__ = ObjectDefinition(  # type: ignore generalTypeIssues
+            name=to_pascal_case(cls.__name__),
```

Comment:
nit: hopefully this isn't needed anymore (not hurting in meantime though; not a blocker)

Reply:
That's not because of the API, it's to normalize with the implicit object that gets created with the name of the module, just to be "safe". I should add a comment, probably in lots of places.